metadata-ingestion/src/datahub/ingestion/source/s3/source.py

```diff
@@ -737,24 +937,68 @@ def s3_browser(
                     for f in list_folders(
                         bucket_name, f"{folder}", self.source_config.aws_config
                     ):
+                        dirs_to_process = []
                         logger.info(f"Processing folder: {f}")
-                        protocol = ContainerWUCreator.get_protocol(path_spec.include)
-                        dir_to_process = self.get_dir_to_process(
-                            bucket_name=bucket_name,
-                            folder=f + "/",
-                            path_spec=path_spec,
-                            protocol=protocol,
+                        if path_spec.traversal_method == FolderTraversalMethod.ALL:
+                            dirs_to_process.append(f)
+                        else:
+                            if (
+                                path_spec.traversal_method
+                                == FolderTraversalMethod.MIN_MAX
+                                or path_spec.traversal_method
+                                == FolderTraversalMethod.MAX
+                            ):
+                                protocol = ContainerWUCreator.get_protocol(
+                                    path_spec.include
+                                )
+                                dirs_to_process_max = self.get_dir_to_process(
+                                    bucket_name=bucket_name,
+                                    folder=f + "/",
+                                    path_spec=path_spec,
+                                    protocol=protocol,
+                                )
+                                dirs_to_process.append(dirs_to_process_max[0])
+
+                            if (
+                                path_spec.traversal_method
+                                == FolderTraversalMethod.MIN_MAX
+                            ):
+                                dirs_to_process_min = self.get_dir_to_process(
+                                    bucket_name=bucket_name,
+                                    folder=f + "/",
+                                    path_spec=path_spec,
+                                    protocol=protocol,
+                                    min=True,
+                                )
+                                dirs_to_process.append(dirs_to_process_min[0])
```

Comment:
Should we push down traversal mode as max or min_max to `get_dirs_to_process` ? Without that, `list_folders` would be done twice ?